src/persist-client/src/internal/apply.rs

```diff
@@ -156,164 +281,205 @@ where
-                            .lease
-                            .timeout_read
-                            .inc_by(u64::cast_from(expiry_metrics.readers_expired));
-
-                        if let Some(gc) = garbage_collection.as_ref() {
-                            debug!("Assigned gc request: {:?}", gc);
-                        }
-
-                        let maintenance = RoutineMaintenance {
-                            garbage_collection,
-                            write_rollup: state.need_rollup(),
-                        };
-
-                        return Ok((state.seqno(), Ok(work_ret), maintenance));
-                    }
-                    CaSResult::ExpectationMismatch => {
-                        cas_mismatch_metric.0.inc();
-                        state_versions
-                            .fetch_and_update_to_current(state)
-                            .await
-                            .expect("shard codecs should not change");
-                        // Intentionally don't backoff here. It would only make
-                        // starvation issues even worse.
-                        continue;
-                    }
+                metrics
+                    .lease
+                    .timeout_read
+                    .inc_by(u64::cast_from(expiry_metrics.readers_expired));
+
+                if let Some(gc) = garbage_collection.as_ref() {
+                    debug!("Assigned gc request: {:?}", gc);
                 }
+
+                let maintenance = RoutineMaintenance {
+                    garbage_collection,
+                    write_rollup: new_state.need_rollup(),
+                };
+
+                Ok(Ok((
+                    new_state.seqno,
+                    Ok((work_ret, new_state)),
+                    maintenance,
+                )))
             }
-        })
-        .await
+            Ok(CaSResult::ExpectationMismatch) => Err(ExpectationMismatch),
+            Err(err) => Ok(Err(err)),
+        }
+    }
+
+    pub fn update_state(&mut self, new_state: TypedState<K, V, T, D>) {
+        let mut state = self.state.lock().expect("lock poisoned");
+        let seqno_before = state.seqno;
+        state.try_replace_state(new_state);
+        self.cached_state.try_update(&state);
+        assert!(
+            seqno_before <= self.cached_state.seqno,
+            "state seqno regressed: {} vs {}",
+            seqno_before,
+            self.cached_state.seqno
+        );
     }
 
     pub async fn fetch_and_update_state(&mut self) {
-        let seqno_before = self.state.seqno;
+        let seqno_before = self.cached_state.seqno;
         self.state_versions
-            .fetch_and_update_to_current(&mut self.state)
+            .fetch_and_update_to_current(&self.state, seqno_before)
             .await
             .expect("shard codecs should not change");
+        self.cached_state
+            .try_update(&self.state.lock().expect("lock poisoned"));
```

Comment:
having to grab the lock again here seems off to me, maybe fetch_and_update_to_current needs to be reworked a bit?

Reply:
removed cached state